pulsar-io/alluxio/src/main/java/org/apache/pulsar/io/alluxio/sink/AlluxioSink.java

```diff
@@ -0,0 +1,284 @@
+            WriteType writeType;
+            try {
+                writeType = WriteType.valueOf(alluxioSinkConfig.getWriteType().toUpperCase());
+            } catch (IllegalArgumentException e) {
+                throw new IllegalArgumentException("Illegal write type when creating Alluxio files, valid values are: "
+                    + Arrays.asList(WriteType.values()));
+            }
+            createFileOptions.setWriteType(writeType);
+        }
+        fileOutStream = fileSystem.createFile(new AlluxioURI(tmpFilePath), createFileOptions);
+    }
+
+    private void closeAndCommitTmpFile() throws AlluxioException, IOException {
+        // close the tmpFile
+        if (fileOutStream != null) {
+            fileOutStream.close();
+        }
+        // commit the tmpFile
+        String filePrefix = alluxioSinkConfig.getFilePrefix();
+        String fileExtension = alluxioSinkConfig.getFileExtension();
+        String newFile = filePrefix + "-" + System.currentTimeMillis() + fileExtension;
+        String newFilePath = fileDirPath + "/" + newFile;
+        fileSystem.rename(new AlluxioURI(tmpFilePath), new AlluxioURI(newFilePath));
+        fileOutStream = null;
+        tmpFilePath = "";
+        recordsNum = 0;
+        lastRotationTime = System.currentTimeMillis();
+    }
+
+    private void deleteTmpFile() throws AlluxioException, IOException {
+        if (!tmpFilePath.equals("")) {
+            fileSystem.delete(new AlluxioURI(tmpFilePath));
+        }
+    }
+
+    private boolean shouldRotate(long now) {
+        boolean rotated = false;
+        if (recordsNum >= rotationRecordsNum) {
+            rotated = true;
+        } else {
+            if (rotationInterval != -1 && (now - lastRotationTime) >= rotationInterval) {
+                rotated = true;
+            }
+        }
+        return rotated;
+    }
+
+    private static byte[] toByteArray(Object obj) throws IOException {
+        byte[] bytes = null;
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+             ObjectOutputStream oos = new ObjectOutputStream(baos)) {
+            oos.writeObject(obj);
+            oos.flush();
+            bytes = baos.toByteArray();
+        } catch (IOException e) {
+            log.error("Failed to serialize the object.", e);
+            throw e;
+        }
+        return bytes;
+    }
+
+    private static byte[] toBytes(Object obj) throws IOException {
+        byte[] bytes;
+        if (obj instanceof String) {
+            String s = (String) obj;
+            bytes = s.getBytes(StandardCharsets.UTF_8);
+        } else if (obj instanceof byte[]) {
+            bytes = (byte[]) obj;
+        } else {
+            bytes = toByteArray(obj);
+        }
+        return bytes;
+    }
+
+    public KeyValue<String, String> extractKeyValue(Record<GenericObject> record) {
+        String key = record.getKey().orElseGet(() -> String.valueOf(record.getValue()));
```

Comment:
record.getValue() is a GenericObject, that basically is a wrapper over a primitive value + schema information

so here you have to use 
`record.getValue().getNativeObject()`
it will be a `String`, a `byte[ ]`, another primitive type or even a `KeyValue` object if the topic is configured with a KeyValue schema.

if you use the record.getValue().toString() you will not get an useful result (it will be the toString of the GenericObject wrapper)

you can use record.getValue().getSchemaType() to inspect the schema type if you prefer.

Reply:
Thanks for pointing this out, I will spend more time to verify this and refine the unit tests.